plugin/core/main.py

```diff
@@ -0,0 +1,300 @@
+import os
+import subprocess
+
+try:
+    from typing import Any, List, Dict, Tuple, Callable, Optional
+    assert Any and List and Dict and Tuple and Callable and Optional
+except ImportError:
+    pass
+
+import sublime_plugin
+import sublime
+
+from .url import filename_to_uri, uri_to_filename
+from .protocol import (
+    Request, Notification, Range
+)
+from .settings import (
+    ClientConfig, settings, load_settings, unload_settings
+)
+from .logging import debug, exception_log, server_log
+from .rpc import Client
+from .workspace import get_project_path
+from .configurations import (
+    config_for_scope, is_supported_view
+)
+from .clients import (
+    add_window_client, window_clients, unload_old_clients,
+    unload_window_clients, unload_all_clients
+)
+from .events import Events
+from .documents import (
+    initialize_document_sync, notify_did_open
+)
+from .diagnostics import handle_diagnostics, remove_diagnostics
+
+
+def plugin_loaded():
```

Comment:
`plugin_loaded()` and `plugin_unloaded()` should be part of the top-level __init__.py. Otherwise reloading the plugin with AutomaticPackageReloader doesn't work. These functions placed here are called only once upon ST startup.

Reply:
Good to know, I will try this out and add instructions so people know to use it.

Reply:
I made the changes and reloading appears to work (hovers still work after reloading the package, at least). But can you explain why these hooks living in `main.py` would prevent reloading from working?

Reply:
It's the way ST (re)loads its plugins. The reason is basically the same as for not reloading modules from sub directories.

It is designed to load top-level modules and look for `plugin_loaded` within their `__dict__` object. The way those modules are reloaded and the imports are handled then, causes this `__dict__` object to not contain `plugin_loaded`.
see: sublime_plugin.py L 101ff.

I guess it is related with the existing "issues" with reloading python modules. `__dict__` is an attribute, which are not or only partly updated by the `imp.reload()` or `importlib.import_module`.

But I am not too much of an python expert to give a real technical explanation for it.